spring-kafka/src/test/java/org/springframework/kafka/retrytopic/AsyncMonoRetryTopicScenarioTests.java

```diff
@@ -0,0 +1,1178 @@
+			q.add(msg);
+		}
+
+		int expectedSuccessMsgCount = 50;
+		int expectedFailedMsgCount = 250;
+
+		int expectedReceivedOriginalTopicCount = 100;
+		int expectedReceivedRetryTopicCount = 200;
+		int expectedReceivedDltMsgCount = 50;
+
+
+		// When
+		while (!q.isEmpty()) {
+			String successOrFailMsg = q.poll();
+			kafkaTemplate.send(TEST_TOPIC6, successOrFailMsg);
+		}
+
+		// Then
+		assertThat(awaitLatch(latchContainer.countDownLatch6)).isTrue();
+		assertThat(awaitLatch(latchContainer.dltCountdownLatch6)).isTrue();
+
+		assertThat(destinationTopic.getDestinationName()).isEqualTo(expectedRetryTopic);
+
+		long actualReceivedSuccessMsgCount = topicListener6.receivedMsgs.stream()
+												.map(s -> s.split(",")[1])
+												.filter(m -> (',' + m).equals(TestTopicListener6.SUCCESS_SUFFIX))
+												.count();
+
+		long actualReceivedFailedMsgCount = topicListener6.receivedMsgs.stream()
+																		.map(s -> s.split(",")[1])
+																		.filter(m -> (',' + m).equals(
+																			TestTopicListener6.FAIL_SUFFIX))
+																		.count();
+
+
+		long actualReceivedOriginalTopicMsgCount = topicListener6.receivedTopics.stream()
+												.filter(topic -> topic.equals(TEST_TOPIC6))
+												.count();
+
+		long actualReceivedRetryTopicMsgCount = topicListener6.receivedTopics.stream()
+												.filter(topic -> topic.equals(expectedRetryTopic))
+												.count();
+
+		assertThat(actualReceivedSuccessMsgCount).isEqualTo(expectedSuccessMsgCount);
+		assertThat(actualReceivedFailedMsgCount).isEqualTo(expectedFailedMsgCount);
+		assertThat(actualReceivedOriginalTopicMsgCount).isEqualTo(expectedReceivedOriginalTopicCount);
+		assertThat(actualReceivedRetryTopicMsgCount).isEqualTo(expectedReceivedRetryTopicCount);
+
+		assertThat(myCustomDltProcessor6.receivedMsg.size()).isEqualTo(expectedReceivedDltMsgCount);
+	}
+
+	private boolean awaitLatch(CountDownLatch latch) {
+		try {
+			return latch.await(60, TimeUnit.SECONDS);
+		}
+		catch (Exception e) {
+			fail(e.getMessage());
+			throw new RuntimeException(e);
+		}
+	}
+
+	@KafkaListener(
+			id = "0-topicId",
+			topics = TEST_TOPIC0,
+			containerFactory = MAIN_TOPIC_CONTAINER_FACTORY,
+			errorHandler = "myCustomErrorHandler",
+			contentTypeConverter = "myCustomMessageConverter",
+			concurrency = "2")
+	static class TestTopicListener0 {
+
+		@Autowired
+		CountDownLatchContainer container;
+
+		private final List<String> receivedMsgs = new ArrayList<>();
+
+		private final List<String> receivedTopics = new ArrayList<>();
+
+		@KafkaHandler
+		public Mono<Void> listen(String message, @Header(KafkaHeaders.RECEIVED_TOPIC) String receivedTopic) {
+			this.receivedMsgs.add(message);
+			this.receivedTopics.add(receivedTopic);
+			return Mono.fromCallable(() -> {
+				container.countDownLatch0.countDown();
+				throw new RuntimeException("Woooops... in topic " + receivedTopic);
```

Comment:
This is still not OK.
The `countDown()` has to happen after we throw an exception.
So, it has to go to the `finally`.